src/mlpack/core/optimizers/fw/frank_wolfe.hpp

```diff
@@ -0,0 +1,177 @@
+#include "func_sq.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+/**
+ * Frank-Wolfe is a technique to minimize a continuously differentiable convex
+ * function \f$ f \f$ over a compact convex subset \f$ D \f$ of a vector space. 
+ * It is also known as conditional gradient method.
+ *
+ * To find minimum of a function using Frank-Wolfe in each iteration \f$ k \f$:
+ * 1. One optimize the linearized constrained problem, using LinearConstrSolver:
+ * \f[
+ * s_k:= arg\min_{s\in D} <s_k, \nabla f(x_k)>
+ * \f]
+ *
+ * 2. Update \f$ x \f$ using UpdateRule:
+ * \f[
+ * x_{k+1} := (1-\gamma) x_k + \gamma s_k
+ * \f]
+ * for some \f$ \gamma \in (0, 1) \f$, or use Fully-Corrective Variant:
+ * \f[
+ * x_{k+1}:= arg\min_{x\in conv(s_0, \cdots, s_k)} f(x)
+ * \f]
+ *
+ *
+ * The algorithm continues until \f$ k \f$ reaches the maximum number of iterations, 
+ * or when the duality gap is bounded by a certain tolerance \f$ \epsilon \f$. 
+ * That is,
+ *
+ * \f[
+ * g(x):= \max_{s\in D} <x-s, \nabla f(x)> \quad \leq \epsilon,
+ * \f]
+ *
+ * we also know that \f$ g(x) \geq f(x) - f(x^*) \f$, where \f$ x^* \f$ is the optimal 
+ * solution.
+ *
+ * The parameter \f$ \epsilon \f$ is specified by the tolerance parameter to the
+ * constructor.
+ *
+ * For FrankWolfe to work, FunctionType, LinearConstrSolverType and UpdateRuleType 
+ * template parameters are required.
+ * These classes must implement the following functions:
+ *
+ * FunctionType:
+ *
+ *   double Evaluate(const arma::mat& coordinates);
+ *   void Gradient(const arma::mat& coordinates,
+ *                 arma::mat& gradient);
+ *
+ * LinearConstrSolverType:
+ *
+ *   void Optimize(const arma::mat& gradient,
+ *                 arma::mat& s);
+ *
+ * UpdateRuleType:
+ *
+ *   void Update(const arma::mat& old_coords,
+ *               const arma::mat& s,
+ *               arma::mat& new_coords,
+ *               const size_t num_iter);
+ *
+ * @tparam FunctionType Objective function type to be
+ *     minimized.
+ * @tparam LinearConstrSolverType Solver for the linear constrained problem.
+ * @tparam UpdateRuleType Rule to update the solution in each iteration.
+ *
+ */
+template<
+    typename FunctionType,
+    typename LinearConstrSolverType,
+    typename UpdateRuleType
+>
+class FrankWolfe
+{
+ public:
+  /**
+   * Construct the Frank-Wolfe optimizer with the given function and
+   * parameters. Notice that the constraint domain \f$ D \f$ is input
+   * at the initialization of linear_constr_solver, the function to be  
+   * optimized is stored in update_rule.
+   *
+   * @param function Function to be optimized
+   * @param linear_constr_solver Solver for linear constrained problem.
+   * @param update_rule Rule for updating solution in each iteration.
+   * @param maxIterations Maximum number of iterations allowed (0 means no
+   *     limit).
+   * @param tolerance Maximum absolute tolerance to terminate algorithm.
+   */
+  FrankWolfe(FunctionType& function,
```

Comment:
#1032, which was recently merged, changes the optimizer API so that the `FunctionType` template parameter is only given to the `Optimize()` function.  Do you think you could make the same change here?  I can clarify what is needed more if you like.

I guess in this situation the constraint domain should also be moved to be input to `Optimize()`.